File: Scripts/deduplicate.py
#!/usr/bin/env python3
"""
Deduplicate and optimize blocklist files
- Removes duplicates within files and across files
- Strips comments, empty lines, and trailing whitespace
- Sorts entries for better compression
- Validates domain syntax
"""
import sys
from pathlib import Path
from collections import defaultdict
from dataclasses import dataclass
from collections.abc import Iterable

# Add current directory to path to allow importing common if run from elsewhere
if str(Path(__file__).parent) not in sys.path:
    sys.path.append(str(Path(__file__).parent))

from common import is_valid_domain, read_lines, write_lines

@dataclass(slots=True)
class Stats:
    original: int = 0
    headers: int = 0
    final: int = 0
    removed: int = 0
  
    @property
    def compression_ratio(self) -> float:
        return (1 - self.final / self.original) * 100 if self.original > 0 else 0.0

def is_header(line: str) -> bool:
    """Check if line is a header/metadata line"""
    return line.startswith(('! ', '#', '[', ';')) or not line

def is_valid_rule(line: str) -> bool:
    """Basic validation for filter rules"""
    if not line or len(line) > 2048:
        return False
    if line.startswith(('||', '@@||')):
        # Extract domain part: remove || or @@||, stop at ^ or $ or options separator
        domain = line.split('^')[0].lstrip('|@')
        return is_valid_domain(domain)
    return True

def process_content(lines: Iterable[str]) -> tuple[list[str], list[str], Stats]:
    """Process lines to separate headers and rules, and deduplicate rules while keeping comments attached."""
    stats = Stats()
    headers = []
    rules_with_comments: list[tuple[str, list[str]]] = []
    seen = set()
    in_header = True
    current_comments = []
  
    for line in lines:
        stats.original += 1
        if not line:
            if in_header:
                headers.append('')
            continue

        if is_header(line):
            if in_header:
                headers.append(line)
            else:
                current_comments.append(line)
        else:
            in_header = False
            if line not in seen and is_valid_rule(line):
                seen.add(line)
                rules_with_comments.append((line, current_comments))
                current_comments = []
            else:
                # Discard comments for duplicate or invalid rules
                current_comments = []

    # Sort rules alphabetically by the rule text
    rules_with_comments.sort(key=lambda x: x[0])

    # Flatten rules and their comments
    rules = []
    for rule, comments in rules_with_comments:
        rules.extend(comments)
        rules.append(rule)

    stats.headers = len(headers)
    stats.final = len(headers) + len(rules)
    stats.removed = stats.original - stats.final

    return headers, rules, stats

def deduplicate_file(filepath: Path) -> tuple[Stats, list[str]]:
    """Deduplicate entries in a single file"""
    print(f"Processing: {filepath}")

    try:
        with filepath.open('r', encoding='utf-8') as f:
            lines_gen = (line.rstrip() for line in f)
            headers, rules, stats = process_content(lines_gen)
    except Exception as e:
        print(f"  Error reading {filepath}: {e}", file=sys.stderr)
        return Stats(), []

    final_content = headers + rules

    if write_lines(filepath, final_content):
        print(f"  {stats.original} → {stats.final} lines ({stats.removed} removed, {stats.compression_ratio:.1f}% reduction)")
        return stats, rules
    
    return Stats(), []

def find_cross_file_duplicates(file_rules: dict[str, list[str]]) -> dict[str, list[str]]:
    """Find entries appearing in multiple files"""
    entry_locations = defaultdict(list)

    for filename, rules in file_rules.items():
        for rule in rules:
            stripped = rule.strip()
            if stripped:
                entry_locations[stripped].append(filename)

    return {entry: files for entry, files in entry_locations.items() if len(files) > 1}

def main() -> int:
    script_dir = Path(__file__).parent
    repo_dir = script_dir.parent

    if len(sys.argv) > 1:
        lists_dir = Path(sys.argv[1])
    else:
        lists_dir = repo_dir / 'lists'
    
    if not lists_dir.exists():
        print(f"Error: Lists directory not found at {lists_dir}", file=sys.stderr)
        return 1

    txt_files = sorted(lists_dir.glob('**/*.txt'))
    if not txt_files:
        print("No .txt files found in lists directory", file=sys.stderr)
        return 1

    print(f"Found {len(txt_files)} files\n")

    total_stats = Stats()
    file_rules = {}
    for filepath in txt_files:
        stats, rules = deduplicate_file(filepath)
        file_rules[filepath.name] = rules
        total_stats.original += stats.original
        total_stats.final += stats.final
        total_stats.removed += stats.removed

    print(f"\n{'='*60}")
    print(f"Total: {total_stats.original} → {total_stats.final} lines")
    print(f"Removed: {total_stats.removed} ({total_stats.compression_ratio:.1f}% reduction)")

    print(f"\n{'='*60}")
    print("Checking for cross-file duplicates...")
    duplicates = find_cross_file_duplicates(file_rules)

    if duplicates:
        file_groups = defaultdict(list)
        for entry, files in duplicates.items():
            file_groups[tuple(sorted(files))].append(entry)

        print(f"Found {len(duplicates)} cross-file duplicates:\n")
        for file_tuple, entries in sorted(file_groups.items()):
            print(f"{len(entries)} entries in: {', '.join(file_tuple)}")
            for entry in sorted(entries)[:5]:
                print(f"  {entry}")
            if len(entries) > 5:
                print(f"  ... and {len(entries) - 5} more")
    else:
        print("✓ No cross-file duplicates")

    return 0

if __name__ == '__main__':
    sys.exit(main())

File: Scripts/deduplicate.ts
#!/usr/bin/env bun
/**
 * Deduplicate and optimize blocklist files
 * - Removes duplicates within files and across files
 * - Strips comments, empty lines, and trailing whitespace
 * - Sorts entries for better compression
 * - Validates domain syntax
 */

import { join, basename, dirname } from "node:path";
import { Glob } from "bun";
import { isValidDomain, writeLines } from "./common.ts";

const HEADER_PREFIXES = ["! ", "#", "[", ";"];

export class Stats {
    original = 0;
    headers = 0;
    final = 0;
    removed = 0;

    get compressionRatio(): number {
        return this.original > 0 ? (1 - this.final / this.original) * 100 : 0.0;
    }
}

export function isHeader(line: string): boolean {
    return !line || HEADER_PREFIXES.some(prefix => line.startsWith(prefix));
}

export function isValidRule(line: string): boolean {
    if (!line || line.length > 2048) {
        return false;
    }
    if (line.startsWith("||") || line.startsWith("@@||")) {
        const domain = line.split("^")[0].replace(/^\|\||^@@\|\|/, "");
        return isValidDomain(domain);
    }
    return true;
}

export function processContent(lines: Iterable<string>): { headers: string[]; rules: string[]; stats: Stats } {
    const stats = new Stats();
    const headers: string[] = [];
    const rulesWithComments: { rule: string; comments: string[] }[] = [];
    const seen = new Set<string>();
    let inHeader = true;
    let currentComments: string[] = [];

    for (const rawLine of lines) {
        stats.original++;
        const line = rawLine.trim();
        if (!line) {
            if (inHeader) {
                headers.push("");
            }
            continue;
        }

        if (isHeader(line)) {
            if (inHeader) {
                headers.push(line);
            } else {
                currentComments.push(line);
            }
        } else {
            inHeader = false;
            if (!seen.has(line) && isValidRule(line)) {
                seen.add(line);
                rulesWithComments.push({ rule: line, comments: currentComments });
                currentComments = [];
            } else {
                currentComments = [];
            }
        }
    }

    rulesWithComments.sort((a, b) => a.rule.localeCompare(b.rule, "en", { sensitivity: "base" }));

    const rules: string[] = [];
    for (const { rule, comments } of rulesWithComments) {
        rules.push(...comments);
        rules.push(rule);
    }

    stats.headers = headers.length;
    stats.final = headers.length + rules.length;
    stats.removed = stats.original - stats.final;

    return { headers, rules, stats };
}

export async function deduplicateFile(filepath: string): Promise<{ stats: Stats; rules: string[] }> {
    console.log(`Processing: ${filepath}`);

    try {
        const file = Bun.file(filepath);
        if (!(await file.exists())) {
            throw new Error(`ENOENT: no such file or directory, open '${filepath}'`);
        }
        const text = await file.text();
        const linesGen = text.split(/\r?\n/).map(line => line.trim());
        const { headers, rules, stats } = processContent(linesGen);

        const finalContent = [...headers, ...rules];
        if (await writeLines(filepath, finalContent)) {
            console.log(`  ${stats.original} → ${stats.final} lines (${stats.removed} removed, ${stats.compressionRatio.toFixed(1)}% reduction)`);
            return { stats, rules };
        }
        return { stats: new Stats(), rules: [] };
    } catch (e: any) {
        console.error(`  Error reading ${filepath}: ${e.message || e}`);
        return { stats: new Stats(), rules: [] };
    }
}

export function findCrossFileDuplicates(fileRules: Record<string, string[]>): Record<string, string[]> {
    const entryLocations: Record<string, string[]> = {};

    for (const [filename, rules] of Object.entries(fileRules)) {
        for (const rawRule of rules) {
            const rule = rawRule.trim();
            if (rule && !isHeader(rule)) { // Ignore comments in cross-file check
                if (!entryLocations[rule]) {
                    entryLocations[rule] = [];
                }
                entryLocations[rule].push(filename);
            }
        }
    }

    const duplicates: Record<string, string[]> = {};
    for (const [entry, files] of Object.entries(entryLocations)) {
        if (files.length > 1) {
            duplicates[entry] = files;
        }
    }
    return duplicates;
}

export async function main(): Promise<number> {
    const repoDir = join(import.meta.dir, "..");
    const args = process.argv.slice(2);
    const listsDir = args.length > 0 ? args[0] : join(repoDir, "lists");

    const listsDirFile = Bun.file(listsDir); // simple check if exists by checking if it throws on readdir? Bun doesn't have exist for dir.
    // Use glob to find files
    const glob = new Glob("**/*.txt");
    const txtFiles: string[] = [];
    try {
        for await (const file of glob.scan({ cwd: listsDir, absolute: true })) {
            txtFiles.push(file);
        }
    } catch (e) {
         console.error(`Error: Lists directory not found at ${listsDir}`);
         return 1;
    }

    txtFiles.sort();

    if (txtFiles.length === 0) {
        console.error("No .txt files found in lists directory");
        return 1;
    }

    console.log(`Found ${txtFiles.length} files\n`);

    const totalStats = new Stats();
    const fileRules: Record<string, string[]> = {};

    for (const filepath of txtFiles) {
        const { stats, rules } = await deduplicateFile(filepath);
        fileRules[basename(filepath)] = rules;
        totalStats.original += stats.original;
        totalStats.final += stats.final;
        totalStats.removed += stats.removed;
    }

    console.log(`\n${'='.repeat(60)}`);
    console.log(`Total: ${totalStats.original} → ${totalStats.final} lines`);
    console.log(`Removed: ${totalStats.removed} (${totalStats.compressionRatio.toFixed(1)}% reduction)`);
    console.log(`\n${'='.repeat(60)}`);
    console.log("Checking for cross-file duplicates...");

    const duplicates = findCrossFileDuplicates(fileRules);

    if (Object.keys(duplicates).length > 0) {
        const fileGroups: Record<string, string[]> = {};
        for (const [entry, files] of Object.entries(duplicates)) {
            const sortedFiles = [...files].sort().join(", ");
            if (!fileGroups[sortedFiles]) fileGroups[sortedFiles] = [];
            fileGroups[sortedFiles].push(entry);
        }

        console.log(`Found ${Object.keys(duplicates).length} cross-file duplicates:\n`);
        const sortedGroups = Object.entries(fileGroups).sort(([a], [b]) => a.localeCompare(b));

        for (const [fileTuple, entries] of sortedGroups) {
            entries.sort();
            console.log(`${entries.length} entries in: ${fileTuple}`);
            for (const entry of entries.slice(0, 5)) {
                console.log(`  ${entry}`);
            }
            if (entries.length > 5) {
                console.log(`  ... and ${entries.length - 5} more`);
            }
        }
    } else {
        console.log("✓ No cross-file duplicates");
    }

    return 0;
}

if (import.meta.main) {
    process.exitCode = await main();
}
